### PR TITLE
add Ontology support

### PR DESCRIPTION
**Pull request by prusnak**
This PR adds Ontology support. There are ported changes from the old trezor-core PR https://github.com/trezor/trezor-core/pull/323

Before merging the [review suggestions by matejcik](https://github.com/trezor/trezor-core/pull/323/commits/54690cfe20e20ff51d58875f18e9c432cfc260f5) need to be worked on

### CODE REVIEW

File: core/src/apps/ontology/__init__.py [hunk 1]
```diff
@@ -1,15 +1,26 @@
...
 
 def boot():
-    wire.add(MessageType.OntologyGetPublicKey, __name__, "get_public_key")
-    wire.add(MessageType.OntologyGetAddress, __name__, "get_address")
-    wire.add(MessageType.OntologySignTransfer, __name__, "sign_transfer")
-    wire.add(MessageType.OntologySignWithdrawOng, __name__, "sign_withdraw_ong")
-    wire.add(MessageType.OntologySignOntIdRegister, __name__, "sign_ont_id_register")
+    ns = [
+        [CURVE, HARDENED | 44, HARDENED | 1024],
+        [CURVE, HARDENED | 44, HARDENED | 888],
```
Comment (prusnak) [inline core/src/apps/ontology/__init__.py]: Where does 888 come from?

File: core/src/apps/ontology/serialize.py [hunk 2]
```diff
@@ -95,7 +94,8 @@ def _serialize_transfer_payload(transfer: OntologyTransfer):
     native_call = build_native_call("transfer", [[struct]], contract)
 
     # 9 is the maximum possible length of the uvarint prefix
-    ret = bytearray(len(native_call) + 9)
+    # ret = bytearray(len(native_call) + 9)
```
Comment (prusnak) [inline core/src/apps/ontology/serialize.py]: Why this change?

Comment (ciny) [reply core/src/apps/ontology/serialize.py]: should've added a comment. with the original code both tests in test_msg_ontology_sign_tx fail. I need to investigate why and fix it properly.

File: common/protob/messages-ontology.proto [hunk 3]
```diff
@@ -0,0 +1,189 @@
+syntax = "proto2";
+
```
Comment (magnumwallet) [inline common/protob/messages-ontology.proto]: Please add here: `package hw.trezor.messages.ontology;`

**Comment by prusnak** [comment]
All tests are green in the CI (as of  f9fa80d)

File: core/src/apps/ontology/sc/builder.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    """
+    if param == -1:
+        writer.write_byte(ret, opcode.PUSHM1)
+    elif param == 0:
+        writer.write_byte(ret, opcode.PUSH0)
+    elif 0 < param < 16:
+        num = opcode.PUSH1 - 1 + param
+        writer.write_byte(ret, num)
+    else:
+        # number encoded as 8 bytes
```
Comment (matejcik) [inline core/src/apps/ontology/sc/builder.py]: perhaps we could simply inline the relevant part of `push_bytes`
(seems to be:
```
writer.write_byte(ret, 8)
writer.write_uint64(ret, param)
```
)

File: core/src/apps/ontology/sc/native_builder.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+    return ret
+
+
+def _write_native_code_script(ret: bytearray, arr: list) -> None:
+    """
+    Writes native code script from supplied data
+    """
+    for val in reversed(arr):
+        if isinstance(val, str):
```
Comment (matejcik) [inline core/src/apps/ontology/sc/native_builder.py]: is this branch used anywhere?
using hex-encoded string internally is highly suspicious

File: core/src/apps/ontology/layout.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        t1 = Text("Attribute:")
+        t1.normal("Name " + attribute.key)
+        t1.normal("Type: " + attribute.type)
+        t1.normal("Value: " + attribute.value)
+        pages.append(t1)
+
+    return await require_confirm(ctx, Paginated(pages), ButtonRequestType.SignTx)
+
+
+def format_amount_ont(value):
```
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/ontology/layout.py]: done

File: core/src/apps/ontology/sc/opcode.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
+# OpCode constants
+PUSH0 = 0x00  # An empty array of bytes is pushed onto the stack.
+PUSHBYTES75 = 0x4B
+PUSHDATA1 = (
+    0x4C
+)  # The next byte contains the number of bytes to be pushed onto the stack.
```
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/ontology/sc/opcode.py]: done

File: core/src/apps/ontology/serialize.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    if attributes is not None:
+        for attribute in attributes:
+            _serialize_tx_attribute(hw, attribute)
+
+
+def serialize_transfer(transfer: OntologyTransfer):
+    """
+    Serializes transaction with specified transfer as payload
+    """
+    return _serialize_transfer_payload(transfer)
```
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/ontology/serialize.py]: done

File: core/src/apps/ontology/sign.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
+from trezor.crypto.curve import nist256p1
```
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/ontology/sign.py]: Due to protobuf refactor this no longer applies

File: python/trezorctl [hunk 10]
```diff
@@ -1835,6 +1836,195 @@ def tezos_sign_tx(connect, address, file):
...
+
+@cli.command(help="Sign Ontology transfer.")
+@click.option(
+    "-n",
+    "--address",
+    required=True,
+    help="BIP-32 path to signing key, e.g. m/44'/888'/0'/0/0",
+)
+@click.option(
+    "-tx",
```
[the remark this answers is not in the document]
Comment (ciny) [reply python/trezorctl]: fixed all instances of multi-letter short options in ontology commands.

File: core/src/apps/ontology/writer.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
+from apps.wallet.sign_tx.writers import write_varint
```
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/ontology/writer.py]: I imported the relevant functions from apps.common.writers instead. is that okay or should those functions be imported where they are used?

**Comment by tsusanka** [comment]
Travis should be passing after rebase.

**Comment by ciny** [comment]
> Travis should be passing after rebase.

yup, travis passing now.

File: core/src/apps/ontology/serialize.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        contract = Const.ONG_CONTRACT
+
+    struct = ParamStruct([from_address, to_address, amount])
+    native_call = build_native_call("transfer", [[struct]], contract)
+
+    # 9 is the maximum possible length of the uvarint prefix
+    # ret = bytearray(len(native_call) + 9)
+    ret = bytearray()
+    writer.write_bytes_with_length(ret, native_call)
+    return bytes(ret)
```
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/ontology/serialize.py]: I tried to do this but tests stop passing. I suspect it's because the writer used in serialize_tx is trezor.utils.HashWriter. I can still return the native_call and move L84-85 to serialize_tx, does that make sense?

File: core/src/apps/ontology/sc/native_builder.py (hunk 5 above)
Comment (ciny) [reply core/src/apps/ontology/sc/native_builder.py]: it doesn't seem to be, at least not in the device tests

Comment (ciny) [reply core/src/apps/ontology/sc/native_builder.py]: Should I remove the branch?

File: core/src/apps/ontology/writer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/ontology/writer.py]: yes, this is OK

File: core/src/apps/ontology/serialize.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/ontology/serialize.py]: after discussion we determined that we need to keep this here as-is

(it is because the returned payload isn't only used in `serialize_tx`, but also sent back to the host as the `payload` attribute.

OTOH maybe the host doesn't actually need that?)

**Comment by matejcik** [comment]
@magnumwallet are you planning Trezor integration with Ontology?

will you be using the `payload` content returned from Trezor, or will you be serializing the transaction separately and only using the `signature`?

**Comment by magnumwallet** [comment]
> @magnumwallet are you planning Trezor integration with Ontology?
> 
> will you be using the `payload` content returned from Trezor, or will you be serializing the transaction separately and only using the `signature`?

I guess we can only use the signature.

**Comment by tsusanka** [comment]
@ciny any chance we might finish this for the freeze, that is Monday?

File: core/src/apps/ontology/serialize.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/ontology/serialize.py]: After further discussion we decided to drop the payload attribute from protobuf and refactor this as discussed earlier.

File: core/src/apps/ontology/sign_ont_id_register.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+from .layout import require_confirm_ont_id_register
+from .serialize import serialize_ont_id_register, serialize_tx
+from .sign import sign
+
+from apps.common import paths
+
+
+async def sign_ont_id_register(ctx, msg: OntologySignOntIdRegister, keychain):
+    await paths.validate_path(ctx, validate_full_path, keychain, msg.address_n, CURVE)
+    if msg.transaction.type == 0xD1:
```
[the remark this answers is not in the document]
Comment (ciny) [reply core/src/apps/ontology/sign_ont_id_register.py]: What would happen if we receive a transaction of different type for whatever reason (client error or something)?

[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/ontology/sign_ont_id_register.py]: a `wire.DataError` apparently :)

Or are you asking what I propose should happen? My idea is to remove the field, so the client *can't* send a different transaction type.

Or IOW, do we want to support different transaction types, and if so, do we want to only implement this one now? What even *are* different transaction types?

**Comment by tsusanka** [comment]
We are interested in adding this to our firmware but we just don't have enough capacity to review it right now. We would be more than happy to take a closer look at it later, when our capacity frees up a little, likely during the course of 2021. Thank you for your understanding.

**Comment by matejcik** [comment]
to expand: this has gone through several rounds of review and there is still one outstanding issue, namely the complexity of `apps.ontology.sc` module, which seems completely unwarranted.
However, we lack the manpower and expertise in Ontology to resolve this properly. Contributions from the Ontology community would be welcome.
